lua/hbac/telescope/init.lua

```diff
@@ -0,0 +1,119 @@
+if not pcall(require, "telescope") then
+	return false
+end
+
+local M = {}
+
+local make_display = require("hbac.telescope.make_display")
+local utils = require("hbac.utils")
+
+local action_state = require("telescope.actions.state")
+local finders = require("telescope.finders")
+local pickers = require("telescope.pickers")
+local telescope_conf = require("telescope.config").values
+
+local display = make_display.display
+
+local get_entries = function()
+	local entries = {}
+	local buflist = utils.get_buffers()
+	for _, bufnr in ipairs(buflist) do
+		local bufname = vim.api.nvim_buf_get_name(bufnr)
+		table.insert(entries, {
+			display = display,
+			value = bufnr,
+			ordinal = bufname,
+		})
+	end
+	return entries
+end
+
+local make_finder = function()
+	return finders.new_table({
+		results = get_entries(),
+		entry_maker = function(entry)
+			return {
+				value = entry.value,
+				display = entry.display,
+				ordinal = entry.ordinal,
+				path = entry.ordinal,
+			}
+		end,
+	})
+end
+
+local attach_mappings = function(_, map)
+	local actions = require("telescope.actions")
+	local hbac_config = require("hbac.setup").opts
+	local state = require("hbac.state")
+	local subcommands = require("hbac.command.subcommands")
+
+	local execute_telescope_action = function(prompt_bufnr, action)
+		local picker = action_state.get_current_picker(prompt_bufnr)
+		local multi_selection = picker:get_multi_selection()
+
+		if next(multi_selection) then
+			for _, entry in ipairs(multi_selection) do
+				action(entry.value)
+			end
+		else
+			local single_selection = action_state.get_selected_entry()
+			action(single_selection.value)
+		end
+		picker:refresh(make_finder(), { reset_prompt = false })
+	end
+
+
+	local hbac_toggle_selections = function(prompt_bufnr)
+		execute_telescope_action(prompt_bufnr, state.toggle_pin)
+	end
+	local hbac_toggle_all = function(prompt_bufnr)
+		execute_telescope_action(prompt_bufnr, subcommands.toggle_all)
+	end
+	local hbac_close_unpinned = function(prompt_bufnr)
+		execute_telescope_action(prompt_bufnr, subcommands.close_unpinned)
+	end
+	local hbac_delete_buffer = function(prompt_bufnr)
+		execute_telescope_action(prompt_bufnr, hbac_config.close_command)
+	end
+
+	actions.hbac_toggle_selections = hbac_toggle_selections
+	actions.hbac_toggle_all = hbac_toggle_all
+	actions.hbac_close_unpinned = hbac_close_unpinned
+	actions.hbac_delete_buffer = hbac_delete_buffer
```

Comment:
To show up in the Telescope which_key popup, the functions need to be registered as named functions in `telescope.actions`, and previously I was just mapping them in a loop. This is kind of clunky but I'm having a hard time coming up with anything cleaner right now due to the named-function constraint. Need to stop staring at it for a while and maybe something will come. Appreciate any suggestions.

Reply:
@ditsuke I tried this initially but this registers them as anonymous functions so it goes back to displaying
<img src=https://user-images.githubusercontent.com/32141102/238931358-a9f43989-cacc-4eba-8c65-46daee52b22a.png>

I want to dig through the telescope code a little more so I can understand this, but so far I've only been able to get the which_key pop up hints to register with the same name they're assigned in the attach_mappings function. The `map()` function doesn't care what it gets, and that's why we can iterate through the mappings opt to set those. I'll keep looking though, I'm still not understanding the fundamentals of what's going on here.

Reply:
So this won't work right now? I'm getting the <anonymous> still. I guess we could create an issue and leave it for later, if you want.

Reply:
@axkirillov sorry, something to do with the modularization of `attach_mappings` I did in 90689bec47bf24c782c9b1cd443ebf3bfd353825, will revert and go from there

Reply:
Pushed [28fd951](https://github.com/axkirillov/hbac.nvim/pull/5/commits/28fd9513918af49fd555eceef4ac4bd6e5581ea9), how about now?

Telescope's [developers.md](https://github.com/nvim-telescope/telescope.nvim/blob/master/developers.md) mentions:

> We [replace the default action] by setting the attach_mappings key to a function. This function needs to return either true or false. If it returns false it means that only the actions defined in the function should be attached.

So I guess if I want to clean this up I need to figure out the constraints of "defined in the function". I'm also ok with leaving it as a task for someone more skilled :sweat_smile:

Reply:
It works now 👍🏻

Reply:
Works! :100: